EIPS/eip-6492.md

```diff
@@ -0,0 +1,143 @@
+
+Note that we're passing `factoryCalldata` instead of `salt` and `bytecode`. We do this in order to make verification compliant with any factory interface. We do not need to calculate the address based on  `create2Factory`/`salt`/`bytecode`, because [ERC-1271](./eip-1271.md) verification presumes we already know the account address we're verifying the signature for.
+
+### Verifier side
+
+Full signature verification MUST be performed in the following order:
+
+- check if there's contract code at the address. If so perform [ERC-1271](./eip-1271.md) verification as usual by invoking `isValidSignature`
+- check if the signature ends with magic bytes, in which case do an `eth_call` to a multicall contract that will call the factory first with the `factoryCalldata` and deploy the contract; Then, call `contract.isValidSignature` as usual
+- If all this fails, try `ecrecover` verification
+
+## Rationale
+
+We believe that wrapping the signature in a way that allows to pass the deploy data is the only clean way to implement this, as it's completely contract agnostic, but also easy to verify.
+
+The wrapper format ends in `magicBytes`, which ends with a `0x69`, which makes it is impossible for it to collide with a valid `ecrecover` signature if packed in the `r,s,v` format, as `0x69` is not a valid value for `v`.
+
+The verification order to ensure correct verification is based on the following rules:
+
+- checking `ecrecover` MUST NOT happen before [ERC-1271](./eip-1271.md) verification, because a contract may use a signature format that also happens to be a valid `ecrecover` signature for an EOA with a different address. One such example is a contract that's a wallet controlled by said EOA.
+- checking for `magicBytes` MUST happen before `ecrecover` in order to avoid trying to verify a counterfactual contract signature via `ecrecover` if such is clearly identifiable
+- checking for contract code MUST happen before checking for `magicBytes`, to avoid possible collisions, as [ERC-1271](./eip-1271.md) allows any arbitrary values for signatures
+
+## Backwards Compatibility
+
+This ERC is backward compatible with previous work on signature validation, including [ERC-1271](./eip-1271.md) and allows for easy verification of all signature types, including EOA signatures and typed data ([EIP-712](./eip-712.md)). 
+
+## Reference Implementation
+
+Example implementation of a verification contract:
+
+```solidity
+contract ERC6492Verifier {
+  /**
+   * @notice Verifies that the signature is valid for that signer and hash
+   */  
+  function isValidUniversalSignature(
+    address _signer,
+    bytes32 _hash,
+    bytes calldata _signature
+  ) external override returns (bool) {
+    uint32 contractSize;
+    assembly {
+      size := extcodesize(_signer)
+    }
+    if (contractSize > 0) {
+      return IERC1271Wallet(_signer).isValidSignature(_hash, _signature) == 0x1626ba7e;
```

Comment:
this logic would fail if the contract gets deployed between the time the signature was created, and the time it was checked.
The magic check should be first, and the deploy code skipped, in case the contract is already deployed.

Reply:
This is a matter of compromise as this would allow collisions, but it's a good point. If addressed, the magic value will probably need to be bytes32 though to reduce the chance of collisions.